src/main/java/scala/compat/java8/runtime/LambdaDeserializer.scala

```diff
@@ -0,0 +1,135 @@
+package scala.compat.java8.runtime
+
+import java.lang.invoke._
+import java.lang.ref.WeakReference
+
+/**
+ * This class is only intended to be called by synthetic `$deserializeLambda$` method that the Scala 2.12
+ * compiler will add to classes hosting lambdas.
+ *
+ * It is intended to be consumed directly.
+ */
+object LambdaDeserializer {
+  private final case class CacheKey(implClass: Class[_], implMethodName: String, implMethodSignature: String)
+  private val cache = new java.util.WeakHashMap[CacheKey, WeakReference[CallSite]]()
```

Comment:
Or maybe we don't need weak values. I'm going to try to make a test that loads a lambda hosting class in a nested classloader. I'll then create a lambda hosted by that class, serialize and deserialize it, drop the reference to the classloader, churn GC, and see if the cache entry clears. My worry is that the strong reference to the lambda class will prevent GC of its containing classloader.